keps/0020-manual-plan-execution.md

```diff
@@ -0,0 +1,175 @@
+---
+kep-number: 20
+title: Manual plan execution
+authors:
+  - "@alenkacz"
+owners:
+  - "@alenkacz"
+creation-date: 2019-11-08
+last-updated: 2019-11-08
+status: provisional
+see-also:
+  - KEP-18
+---
+
+# manual-plan-execution
+
+## Table of Contents
+
+   * [manual-plan-execution](#manual-plan-execution)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Update the status of Instance with name of the plan you want to run, let controller pick it up](#update-the-status-of-instance-with-name-of-the-plan-you-want-to-run-let-controller-pick-it-up)
+            * [Pros](#pros)
+            * [Cons](#cons)
+         * [Expose HTTP API from controller to allow receiving commands](#expose-http-api-from-controller-to-allow-receiving-commands)
+            * [Pros](#pros-1)
+            * [Cons](#cons-1)
+         * [Have a separate CRD just for maintaining commands to execute a plan](#have-a-separate-crd-just-for-maintaining-commands-to-execute-a-plan)
+            * [Pros](#pros-2)
+            * [Cons](#cons-2)
+         * [Introduce new sub-resource for triggering plan executions](#introduce-new-sub-resource-for-triggering-plan-executions)
+            * [Pros](#pros-3)
+            * [Cons](#cons-3)
+         * [User Stories](#user-stories)
+            * [As an operator (person) I want to be able to run a custom plan on my KUDO operator in order to trigger non-default plan](#as-an-operator-person-i-want-to-be-able-to-run-a-custom-plan-on-my-kudo-operator-in-order-to-trigger-non-default-plan)
+            * [As a stateful service operator (person) I want to be able to run backup on my stateful service when a backup plan is defined.](#as-a-stateful-service-operator-person-i-want-to-be-able-to-run-backup-on-my-stateful-service-when-a-backup-plan-is-defined)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Since the inception of KUDO, being able to manually run a plan was one of the core features. After KEP-18 (controller overhaul) we changed the way we store information about executed plans practically removing the ability to execute custom plan in an easy way through kubectl. Goal of this KEP is to propose a way how to re-introduce that functionality and provide a good user experience.
+
+## Motivation
+
+### Goals
+
+Operator will be able to trigger plan of given name through CLI
+Controller won’t allow multiple plans running in parallel
+
```

Comment:
1. it's a bit odd to have a negative as a goal.   "controller won't allow" should this be a Non-Goal?
2. It should be added that kudo will not queue up triggers (plan for foo is running and someone triggers bar.. that doesn't queue bar... if hopefully errors with a non-zero exit on cli.

Reply:
That's actually an interesting question - do we reject such request or do we accept it and then mark as rejected. I think we cannot really guarantee that we'll always be able to reject because there are multiple things happening in parallel that are not synchronized:
- user adding new CRD
- controller picking up on existing things and starting new plan

I think this scenario is totally relevant and could not be prevented via webhook:
- user submits a new ExecutionRequest, that is accepted, at this point, no plan is in progress
- another user submits update to the instance, triggering some kind of plan
- reconciliation is triggered and controller picks up on both instance change and ExecutionRequest

We'll need to think about these corner cases. Overall I would say - we should try to reject in webhook if possible, and plus on top of that handle these racy situations inside controller